### PR TITLE
dns: fix lifetime issue with local context

### PR DESCRIPTION
**Pull request by alyssawilk**
Make sure the DNS code can handle the creating factory (and context) being destroyed before the DNS cache.

Risk Level: low
Testing: new unit tests
Docs Changes: n/a 
Release Notes: n/a

### CODE REVIEW

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Awesome, thanks for the quick fix. Do you mind unreverting https://github.com/envoyproxy/envoy/commit/0656e34e0c87a77fb87273be89afd4922a30c7db#diff-d5ea5a7de19bb4726ab21bb51687efa3526b35914af2af1a8544e50fcccb45c5, and then calling getCache(...) in one of the DFP integration tests? That would have crashed previously. It should be pretty trivial to add, you will just need to post the call to the main thread. Thank you!

/wait

**Comment by alyssawilk** [comment]
I'm not quite sure what you mean regarding testing.  Do you mean the client thread should try to get a dns cache, or we should have a filter which posts a getCache attempt on the main server thread?  Either way I'm not sure it'd repro the problem since the problem is AFIK the factory context going out of scope too early, and I don't think that'd happen for Envoy-the-server.  I've added a unit test to specifically test the lifetime issue and confirmed it fails on asan.  I'm inclined to think if we want to integration test lifetime, we should do it on the envoy mobile side since that has the threading and factory set up to reproduce the issue?

**Comment by mattklein123** [comment]
> Do you mean the client thread should try to get a dns cache, or we should have a filter which posts a getCache attempt on the main server thread?

I was suggesting having the test thread (not a filter) post a getCache() attempt to the main thread and make sure it returns a cache. IMO that would be valuable, but if your unit test crashes also before the fix that's fine with me.

File: bazel/envoy_internal.bzl [hunk 1]
```diff
@@ -68,7 +68,7 @@ def envoy_copts(repository, test = False):
                    "-Wc++2a-extensions",
                    "-Wrange-loop-analysis",
                ],
-               repository + "//bazel:gcc_build": ["-Wno-maybe-uninitialized"],
```
Comment (mattklein123) [inline bazel/envoy_internal.bzl:71]: ?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small merge issue I think.

/wait

**Review by mattklein123 — APPROVED** [review]

**Comment by alyssawilk** [comment]
cc @goaway @buildbreaker @mergeconflict sorry for the issue but this should fix. lmk if you want me to bump envoy-mobile to pick up the fix

**Pull request merged**